src/interpreter.rs

```diff
@@ -1,222 +1,417 @@
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-                            let lhs = match lhs_value {
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-
-                            if !rhs.contains(lhs) {
+                    }
+                }
+                BinaryOperator::In => match (lhs_value, &self.rhs) {
+                    (Value::String(l), Value::String(r)) => {
+                        if r.contains(l) {
+                            curr_match = true;
+                            if any {
                                 return true;
                             }
                         }
-                        BinaryOperator::Contains => {
-                            let rhs = match &p.rhs {
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-                            let lhs = match lhs_value {
-                                Value::String(s) => s,
-                                _ => unreachable!(),
-                            };
-
-                            if lhs.contains(rhs) {
+                    }
+                    (Value::IpAddr(l), Value::IpCidr(r)) => {
+                        if r.contains(l) {
+                            curr_match = true;
+                            if any {
                                 return true;
                             }
                         }
                     }
-                }
+                    _ => unreachable!(),
+                },
+                BinaryOperator::NotIn => {
+                    let rhs = match &self.rhs {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+                    let lhs = match lhs_value {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+
+                    if !rhs.contains(lhs) {
+                        curr_match = true;
+                        if any {
+                            return true;
+                        }
+                    }
+                },
+                BinaryOperator::Contains => {
+                    let rhs = match &self.rhs {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
+                    let lhs = match lhs_value {
+                        Value::String(s) => s,
+                        _ => unreachable!(),
+                    };
 
-                // no match for all values
-                false
+                    if lhs.contains(rhs) {
+                        return true;
+                    }
+                }
+            } // match
+            if !any && !curr_match {
```

Comment:
```suggestion

            if !any && !curr_match {
```

Style: add a newline before and after `if` blocks.